### PR TITLE
TMDM-10425 Be able to filter a foreign key using entity fields

### PR DESCRIPTION
**Pull request by zfkang-talend**
JIRA: https://jira.talendforge.org/browse/TMDM-10425
**What is the current behavior?** (You should also link to an open issue here)

**What is the new behavior?**
Add the function to support the XSI function.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java [hunk 1]
```diff
@@ -14,12 +14,31 @@
...
+import com.google.gwt.regexp.shared.MatchResult;
+import com.google.gwt.regexp.shared.RegExp;
+
 /**
  * created by yjli on 2013-8-1 Detailled comment
  *
  */
 public class CommonUtil {
 
+    public static final String FN_PREFIX = "fn:"; //$NON-NLS-1$
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java]: How about change these `public` to `private`?

File: org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java [hunk 2]
```diff
@@ -205,23 +224,44 @@ public static boolean isRelativePath(String foreignKeyFilterValue) {
         return (foreignKeyFilterValue.startsWith(".") || foreignKeyFilterValue.startsWith("..")); //$NON-NLS-1$ //$NON-NLS-2$
     }
 
+    public static boolean isFunction(String foreignKeyFilterValue) {
+        return foreignKeyFilterValue.startsWith(FN_PREFIX);
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java]: What will happen if the input value is null?

File: org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java [hunk 3]
```diff
@@ -264,4 +304,31 @@ public static String unescapeXml(String value) {
...
+    public static Map<String, String> getArgumentsWithXpath(String function){
+        Map<String, String> arguments = new HashMap<String, String>();
+        RegExp reg = RegExp.compile("\\((.*)\\)"); //$NON-NLS-1$
+        MatchResult matchResult = reg.exec(function);
+        String value = EMPTY_STR;
+        while(matchResult != null){
+            value = matchResult.getGroup(0);
+            break;
+        }
+        RegExp regExp = RegExp.compile("((xpath:(([a-zA-Z]*)|((\\.)+)))|/([a-zA-Z]*))/(([a-zA-Z]*)/*)*", "g"); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java]: Better to add an example comment about the RegExp purpose.

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java [hunk 4]
```diff
@@ -391,6 +403,82 @@ public void onSuccess(ItemBean result) {
...
+        if (foreignKeyFilter != null && foreignKeyFilter.contains(CommonUtil.FN_PREFIX)) {
+            ForeignKeyField foreignKeyField = event.getData(BrowseRecords.FOREIGN_KEY_FIELD);
+            String[] criterias = CommonUtil.getCriteriasByForeignKeyFilter(foreignKeyFilter);
+            if (foreignKeyField instanceof ForeignKeySelector) {
+                ForeignKeySelector foreignKeySelector = (ForeignKeySelector) foreignKeyField;
+                for (String cria : criterias) {
+                    Map<String, String> conditionMap = CommonUtil.buildConditionByCriteria(cria);
+                    String filterValue = conditionMap.get(CommonUtil.VALUE_STR);
+                    if (CommonUtil.isFunction(filterValue)) {
+                        if (CommonUtil.containsXPath(filterValue)) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java]: Above two if conditions can be in one line with `&&`?

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsView.java [hunk 5]
```diff
@@ -500,4 +539,31 @@ private void onViewLineageItem(AppEvent event) {
     protected void setDefaultView(final AppEvent event) {
         ItemsToolBar.getInstance().updateEntityCombo();
     }
+
+    private void onTransformFkFilter(AppEvent event) {
+        LinkedList<String> filterValue = new LinkedList<String>((List<String>) event.getData());
+        String foreignKeyFilter = event.getData("foreignKeyFilter");
+        String[] criterias = org.talend.mdm.webapp.base.shared.util.CommonUtil.getCriteriasByForeignKeyFilter(foreignKeyFilter);
+        StringBuilder sb = new StringBuilder();
+        for (String cria : criterias) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsView.java]: Change from `cria` to `criteria`.

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ForeignKey/ForeignKeyField.java [hunk 6]
```diff
@@ -249,11 +256,21 @@ public void setEnabled(boolean enabled) {
...
 
             @Override
             public void onClick(ClickEvent ce) {
                 if (foreignConceptName != null) {
+
+                    String foreignKeyFilter = getOriginForeignKeyFilter();
+                    if (foreignKeyFilter.contains("fn")) {
+                        AppEvent event = new AppEvent(BrowseRecordsEvents.TransformFkFilterItem, foreignKeyFilter);
+                        event.setData(BrowseRecords.FOREIGN_KEY_FIELD, _this);
+                        event.setData("foreignKeyFilter", foreignKeyFilter);
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ForeignKey/ForeignKeyField.java]: Can we change `foreignKeyFilter` to a constant value as the above line?

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ForeignKey/ForeignKeySelector.java [hunk 7]
```diff
@@ -228,72 +226,10 @@ public String parseForeignKeyFilter() {
...
-                                            .unwrapFkValue(targetForeignKeyBean.getId());
-                                } else {
-                                    filterValue = targetNode.getObjectValue().toString();
-                                }
-                            } else {
-                                filterValue = ""; //$NON-NLS-1$
-                            }
-                        }
-                    }
+                    //String[] rightValueOrPathArray = filterValue.split("/"); //$NON-NLS-1$
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ForeignKey/ForeignKeySelector.java]: Can we remove the comment line?

File: org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java (hunk 2 above)
Comment (zfkang-talend) [reply org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java]: in the origin invoke have one judgement.

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsView.java [hunk 8]
```diff
@@ -500,4 +538,31 @@ private void onViewLineageItem(AppEvent event) {
     protected void setDefaultView(final AppEvent event) {
         ItemsToolBar.getInstance().updateEntityCombo();
     }
+
+    private void onTransformFkFilter(AppEvent event) {
```
Comment (lpw82gf) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsView.java]: Comment of the method detail needed.

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/server/ruleengine/DisplayRuleEngine.java [hunk 9]
```diff
@@ -189,18 +191,38 @@ private String getRealTypePath(org.dom4j.Element el) {
...
+                String style = generateFKFilterStyle(xpath, funcitonList);
+                org.dom4j.Document transformedDocumentValue = XMLUtils.styleDocument(dom4jDoc, style);
+                List<org.dom4j.Node> valueNodeList = transformedDocumentValue.selectNodes(xpath);
+                for (org.dom4j.Node valueNode : valueNodeList) {
+                    resultValue.add(valueNode.getText());
+                }
+            }
+
+        } catch (Exception e) {
+            LOG.error(e.getMessage());
```
Comment (lpw82gf) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/server/ruleengine/DisplayRuleEngine.java]: LOG.error("self-defined message", e);

File: org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java (hunk 3 above)
Comment (zfkang-talend) [reply org.talend.mdm.webapp.base/src/main/java/org/talend/mdm/webapp/base/shared/util/CommonUtil.java]: Add the comment in the method head.

**Pull request merged**